### PR TITLE
Add logic to support `pop()` and `remove()` operations for tuples at runtime

### LINKED ISSUE #42954: [Improvement]: Support `remove()` and `pop()` operations for tuples
### Description

Sub task of https://github.com/ballerina-platform/ballerina-lang/issues/41431

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by ravinperera00**
Fixes #42954 

## Approach
Both `remove()` and `pop()` involve removing an element from the tuple. In remove, the removal can be from a selected index while in pop it has to be the last element of the tuple. The definitions for both methods are as follows.

```
# Removes a member of an array.
#
# + arr - the array
# + index - index of member to be removed from `arr`
# + return - the member of `arr` that was at `index`
# This removes the member of `arr` with index `index` and returns it.
# It panics if there is no such member.
public isolated function remove(Type[] arr, int index) returns Type = external;
```
```
# Removes and returns the last member of an array.
# The array must not be empty.
#
# + arr - the array
# + return - removed member
public isolated function pop(Type[] arr) returns Type = external;
```
Since `shift()` involves removing the first element of the tuple, we should be able to tweak and reuse the logic and checks used for that in both `pop` and `remove`. Once an element is removed from a particular location, the members after that location should be moved to fill the empty spot. This rearrangement will have to satisfy inherent type properties of the tuple as discussed in the `shift()` operation (refer https://github.com/ballerina-platform/ballerina-lang/pull/42815). Instead of checking all the members for matching inherent type, we only need to check elements after the removed element.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal [hunk 1]
```diff
@@ -1907,6 +1924,23 @@ function testArrayPop() {
...
     }
+
+    [int, string, int...] tuple1 = [1, "hello", 4];
+    int|string result2 = tuple1.pop();
+    assertTrue(result2 is int);
+    assertValueEquality(4, result2);
+    int|string|error result3 = trap tuple1.pop();
+    assertTrue(result3 is error);
+    if (result3 is error) {
+        assertValueEquality("{ballerina/lang.array}OperationNotSupported", result3.message());
```
Comment (SasinduDilshara) [inline langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal:1935]:
```suggestion
        assertValueEquality("{ballerina/lang.array}OperationNotSupported", result3.message());
```
Is giving `OperationNotSupported` is correct?, we support it for tuples ryt?
Imo, There should be a separate error.  ex:- Invalid operations

File: langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal [hunk 2]
```diff
@@ -1907,6 +1924,23 @@ function testArrayPop() {
...
+        assertValueEquality("the number of members in a tuple value should be greater than the member types of the tuple to perform this operation",
+        <string> checkpanic result3.detail()["message"]);
+    }
+
+    [int, string, Person2...] tuple2 = [1, "hello", {id:5, name:"John", age:56}];
```
Comment (SasinduDilshara) [inline langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal:1940]: Shall we add a test case for `pop()` in empty tuples, Recursive tuples, tuple with type reference type

Comment (ravinperera00) [reply langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal:1940]: I'll add those.

File: langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal (hunk 1 above)
Comment (ravinperera00) [reply langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal:1935]: It was decided in https://github.com/ballerina-platform/ballerina-lang/pull/42815#discussion_r1635091908 to use `OperationNotSupported` for `shift`. Since we are reusing `shift` for `pop` and `remove`, the error is reused.

File: langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Remove.java [hunk 3]
```diff
@@ -18,10 +18,8 @@
 
 package org.ballerinalang.langlib.array;
 
-import io.ballerina.runtime.api.utils.TypeUtils;
```
Comment (HindujaB) [inline langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Remove.java:21]: We can remove unnecessary new lines in the class.

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TupleValueImpl.java [hunk 4]
```diff
@@ -825,20 +825,20 @@ private void resetSize(int index) {
         }
     }
 
-    private void validateTupleSizeAndInherentType() {
+    private void validateTupleSizeAndInherentType(long index) {
```
Comment (HindujaB) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TupleValueImpl.java:828]: Since we cast the index to `int` in multiple places, we can pass the index as `int` and cast it when we call the `validateTupleSizeAndInherentType`.

File: langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Remove.java (hunk 3 above)
Comment (ravinperera00) [reply langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Remove.java:21]: I'll remove them

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TupleValueImpl.java (hunk 4 above)
Comment (ravinperera00) [reply bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/TupleValueImpl.java:828]: I thought we supported large values for indexing in tuples and the casting would make it erroneous. I was wrong. I'll change that. Thank you for pointing that out.

File: langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal [hunk 5]
```diff
@@ -707,7 +707,7 @@ function testShiftOnTupleWithoutValuesForRestParameter() {
     string detailMessage = message is error ? message.toString() : message.toString();
     assertValueEquality("{ballerina/lang.array}OperationNotSupported", err.message());
     assertValueEquality("the number of members in a tuple value should be greater " +
-    "than the member types of the tuple to perform a 'shift' operation", detailMessage);
+    "than the member types of the tuple to perform this operation", detailMessage);
```
Comment (warunalakshitha) [inline langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal:710]: a 'shift' operation looks nice for me than saying this operation.

Comment (ravinperera00) [reply langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal:710]: It was changed because the same function and error was reused in more than one operation such as `shift` and `remove`. Will change the error to display the operation name by passing it as an argument to the function.

Comment (ravinperera00) [reply langlib/langlib-test/src/test/resources/test-src/arraylib_test.bal]: Changes added in 70d6ff420ac8a0265d1361c26cc5b14b88a90b8c and commits proceeding that.

**Review by warunalakshitha — APPROVED** [review]
LGTM
